### PR TITLE
Follow up #36741: improve code structure

### LINKED ISSUE #37905: Doctests failures in sage/doctest/control.py when sagemath-meataxe is installed
### Steps To Reproduce

This is observed in sage-on-gentoo with both python 3.11 and python 3.12 and sage 10.4.beta4. Install sagemath-meataxe and run 
```
sage -t --long /usr/lib/python3.11/site-packages/sage/doctest/control.py
```

### Expected Behavior

When sagemath-meataxe is not installed, the doctest pass
```
sage -t --long --random-seed=309974202432976880116456820555380509718 /usr/lib/python3.11/site-packages/sage/doctest/control.py
    [229 tests, 1.46 s]
----------------------------------------------------------------------
All tests passed!
----------------------------------------------------------------------
```

### Actual Behavior

The doctest fails with
```
sage -t --long --random-seed=170181526529337143612090850685532611974 /usr/lib/python3.11/site-packages/sage/doctest/control.py
**********************************************************************
File "/usr/lib/python3.11/site-packages/sage/doctest/control.py", line 1447, in sage.doctest.control.DocTestController.run
Failed example:
    DC.run()
Expected:
    Running doctests with ID ...
    Using --optional=sage...
    Features to be detected: ...
    Doctesting 1 file.
    sage -t ....py
        [4 tests, ... s]
    ----------------------------------------------------------------------
    All tests passed!
    ----------------------------------------------------------------------
    Total time for all tests: ... seconds
        cpu time: ... seconds
        cumulative wall time: ... seconds
    Features detected...
    0
Got:
    Running doctests with ID 2024-04-30-13-53-30-5803c5d0.
    Running with SAGE_LOCAL='/usr' and SAGE_VENV='/usr'
    Using --optional=sage
    Features to be detected: 4ti2,benzene,bliss,buckygen,conway_polynomials,csdp,cvxopt,cvxopt,database_cremona_ellcurve,database_cremona_mini_ellcurve,database_cubic_hecke,database_ellcurves,database_graphs,database_jones_numfield,database_knotinfo,dvipng,ecm,fpylll,fricas,gap_package_atlasrep,gap_package_design,gap_package_grape,gap_package_guava,gap_package_hap,gap_package_polycyclic,gap_package_qpa,gap_package_quagroup,gfan,graphviz,imagemagick,ipython,jmol,jupymake,kenzo,latte_int,lrcalc_python,lrslib,matroid_database,mcqd,meataxe,mpmath,msolve,nauty,networkx,numpy,palp,pandoc,pdf2svg,pdftocairo,pexpect,phitigra,pillow,plantri,polytopes_db,polytopes_db_4d,pplpy,primecountpy,ptyprocess,pynormaliz,pyparsing,python_igraph,requests,rubiks,sage.combinat,sage.geometry.polyhedron,sage.graphs,sage.groups,sage.libs.braiding,sage.libs.ecl,sage.libs.flint,sage.libs.gap,sage.libs.linbox,sage.libs.m4ri,sage.libs.ntl,sage.libs.pari,sage.libs.singular,sage.misc.cython,sage.modular,sage.modules,sage.numerical.mip,sage.plot,sage.rings.complex_double,sage.rings.finite_rings,sage.rings.function_field,sage.rings.number_field,sage.rings.padics,sage.rings.polynomial.pbori,sage.rings.real_double,sage.rings.real_mpfr,sage.sat,sage.schemes,sage.symbolic,sage_numerical_backends_coin,sagemath_doc_html,scipy,singular,sphinx,symengine_py,sympy,tdlib,threejs
    Doctesting 1 file.
    sage -t --random-seed=0 /tmp/tmp3o72lmtj/tmp_qd1plq6l.py
        [5 tests, 0.12 s]
    ----------------------------------------------------------------------
    All tests passed!
    ----------------------------------------------------------------------
    Total time for all tests: 0.1 seconds
        cpu time: 0.1 seconds
        cumulative wall time: 0.1 seconds
    Features detected for doctesting: meataxe,sage.libs.singular,sage.modules,sage.rings.finite_rings,sage.rings.number_field,sage.rings.real_mpfr,sage.symbolic,sympy
    Features that have been hidden: meataxe
    0
**********************************************************************
File "/usr/lib/python3.11/site-packages/sage/doctest/control.py", line 1465, in sage.doctest.control.DocTestController.run
Failed example:
    DC.run()
Expected:
    Running doctests with ID ...
    Using --optional=sage
    Features to be detected: ...
    Doctesting 1 file.
    sage -t ....py
        [4 tests, ... s]
    ----------------------------------------------------------------------
    All tests passed!
    ----------------------------------------------------------------------
    Total time for all tests: ... seconds
        cpu time: ... seconds
        cumulative wall time: ... seconds
    Features detected...
    0
Got:
    Running doctests with ID 2024-04-30-13-53-30-601ce747.
    Running with SAGE_LOCAL='/usr' and SAGE_VENV='/usr'
    Using --optional=sage
    Features to be detected: 4ti2,bliss,conway_polynomials,csdp,cvxopt,cvxopt,database_cremona_ellcurve,database_cremona_mini_ellcurve,database_cubic_hecke,database_ellcurves,database_graphs,database_jones_numfield,database_knotinfo,dvipng,ecm,fpylll,fricas,gap_package_atlasrep,gap_package_design,gap_package_grape,gap_package_guava,gap_package_hap,gap_package_polycyclic,gap_package_qpa,gap_package_quagroup,gfan,graphviz,imagemagick,ipython,jmol,jupymake,kenzo,latte_int,lrcalc_python,lrslib,matroid_database,mcqd,meataxe,mpmath,msolve,nauty,networkx,numpy,palp,pandoc,pdf2svg,pdftocairo,pexpect,phitigra,pillow,plantri,polytopes_db,polytopes_db_4d,pplpy,primecountpy,ptyprocess,pynormaliz,pyparsing,python_igraph,requests,rubiks,sage.combinat,sage.geometry.polyhedron,sage.graphs,sage.groups,sage.libs.braiding,sage.libs.ecl,sage.libs.flint,sage.libs.gap,sage.libs.linbox,sage.libs.m4ri,sage.libs.ntl,sage.libs.pari,sage.libs.singular,sage.misc.cython,sage.modular,sage.modules,sage.numerical.mip,sage.plot,sage.rings.complex_double,sage.rings.finite_rings,sage.rings.function_field,sage.rings.number_field,sage.rings.padics,sage.rings.polynomial.pbori,sage.rings.real_double,sage.rings.real_mpfr,sage.sat,sage.schemes,sage.symbolic,sage_numerical_backends_coin,sagemath_doc_html,scipy,singular,sphinx,symengine_py,sympy,tdlib,threejs
    Doctesting 1 file.
    sage -t --random-seed=0 /tmp/tmp3o72lmtj/tmp_qd1plq6l.py
        [5 tests, 0.14 s]
    ----------------------------------------------------------------------
    All tests passed!
    ----------------------------------------------------------------------
    Total time for all tests: 0.2 seconds
        cpu time: 0.1 seconds
        cumulative wall time: 0.1 seconds
    Features detected for doctesting: meataxe,sage.libs.singular,sage.modules,sage.rings.finite_rings,sage.rings.number_field,sage.rings.real_mpfr,sage.symbolic,sympy
    Features that have been hidden: meataxe
    0
**********************************************************************
File "/usr/lib/python3.11/site-packages/sage/doctest/control.py", line 1483, in sage.doctest.control.DocTestController.run
Failed example:
    DC.run()                              # optional - meataxe
Expected:
    Running doctests with ID ...
    Using --optional=sage
    Features to be detected: ...
    Doctesting 1 file.
    sage -t ....py
        [4 tests, ... s]
    ----------------------------------------------------------------------
    All tests passed!
    ----------------------------------------------------------------------
    Total time for all tests: ... seconds
        cpu time: ... seconds
        cumulative wall time: ... seconds
    Features detected...
    Features that have been hidden: ...meataxe...
    0
Got:
    Running doctests with ID 2024-04-30-13-53-30-a0fa6d56.
    Running with SAGE_LOCAL='/usr' and SAGE_VENV='/usr'
    Using --optional=sage
    Features to be detected: 4ti2,bliss,conway_polynomials,csdp,cvxopt,cvxopt,database_cremona_ellcurve,database_cremona_mini_ellcurve,database_cubic_hecke,database_ellcurves,database_graphs,database_jones_numfield,database_knotinfo,dvipng,ecm,fpylll,fricas,gap_package_atlasrep,gap_package_design,gap_package_grape,gap_package_guava,gap_package_hap,gap_package_polycyclic,gap_package_qpa,gap_package_quagroup,gfan,graphviz,imagemagick,ipython,jmol,jupymake,kenzo,latte_int,lrcalc_python,lrslib,matroid_database,mcqd,meataxe,mpmath,msolve,nauty,networkx,numpy,palp,pandoc,pdf2svg,pdftocairo,pexpect,phitigra,pillow,plantri,polytopes_db,polytopes_db_4d,pplpy,primecountpy,ptyprocess,pynormaliz,pyparsing,python_igraph,requests,rubiks,sage.combinat,sage.geometry.polyhedron,sage.graphs,sage.groups,sage.libs.braiding,sage.libs.ecl,sage.libs.flint,sage.libs.gap,sage.libs.linbox,sage.libs.m4ri,sage.libs.ntl,sage.libs.pari,sage.libs.singular,sage.misc.cython,sage.modular,sage.modules,sage.numerical.mip,sage.plot,sage.rings.complex_double,sage.rings.finite_rings,sage.rings.function_field,sage.rings.number_field,sage.rings.padics,sage.rings.polynomial.pbori,sage.rings.real_double,sage.rings.real_mpfr,sage.sat,sage.schemes,sage.symbolic,sage_numerical_backends_coin,sagemath_doc_html,scipy,singular,sphinx,symengine_py,sympy,tdlib,threejs
    Doctesting 1 file.
    sage -t --random-seed=0 /tmp/tmp3o72lmtj/tmp_qd1plq6l.py
        [5 tests, 0.12 s]
    ----------------------------------------------------------------------
    All tests passed!
    ----------------------------------------------------------------------
    Total time for all tests: 0.1 seconds
        cpu time: 0.1 seconds
        cumulative wall time: 0.1 seconds
    Features detected for doctesting: meataxe,sage.libs.singular,sage.modules,sage.rings.finite_rings,sage.rings.number_field,sage.rings.real_mpfr,sage.symbolic,sympy
    Features that have been hidden: meataxe
    0
**********************************************************************
1 item had failures:
   3 of  23 in sage.doctest.control.DocTestController.run
    [230 tests, 3 failures, 1.64 s]
----------------------------------------------------------------------
sage -t --long --random-seed=170181526529337143612090850685532611974 /usr/lib/python3.11/site-packages/sage/doctest/control.py  # 3 doctests failed
```

### Additional Information

As can be seen in the failing doctest there is a mention that meataxe is a hidden feature. But in actual fact it is not during this test.

### Environment

```markdown
- Gentoo
- 10.4.beta4 (sage-on-gentoo)
```

### Checklist

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.
- [X] I have read the documentation and troubleshoot guide

### PR DESCRIPTION
**Pull request by soehms**
From https://github.com/sagemath/sage/pull/36741#issuecomment-2027705010

> Overall, I would suggest to remove the direct and unconditional use of multiprocessing from sage.features.
> Perhaps sage.doctest can put such Value attributes into features that are to be hidden.

This PR implements the suggestion made in https://github.com/sagemath/sage/pull/36741#issuecomment-2028145753

> I looked a little bit into it and here's an idea:
> 
> * In features, implement a simple hide() / unhide() / is_hidden() interface.
> * The only state is _hidden, not shared (for parallel doctesting this will be set before fork so it's ok)
> * Implement AvailableSoftware.hidden() similar to AvailableSoftware.seen() so the logic stays in that class and internals don't leak to sage.doctest.control (if necessary use _seen[idx] to record hidden state and/or number of hidings, or add another shared array)

Fixes #37905

### :memo: Checklist

- [ ] The title is concise and informative.
- [ ] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by mkoeppe — APPROVED** [review]
LGTM, thanks

**Comment by soehms** [comment]
> LGTM, thanks

Thanks, too!

**Comment by vbraun** [comment]
I'm getting this in parallel tests, but passes in isolation:
```
sage -t --long --warn-long 43.9 --random-seed=123 src/sage/doctest/external.py
**********************************************************************
File "src/sage/doctest/external.py", line 518, in sage.doctest.external.AvailableSoftware.hidden
Failed example:
    available_software.hidden()
Expected:
    [...'conway_polynomials',
     'database_cremona_mini_ellcurve',
     'database_ellcurves',
     'database_graphs'...]
Got:
    ['database_cremona_mini_ellcurve', 'database_ellcurves', 'database_graphs']
**********************************************************************
```

**Comment by tornaria** [comment]
> I'm getting this in parallel tests, but passes in isolation:
> 
> ```
> sage -t --long --warn-long 43.9 --random-seed=123 src/sage/doctest/external.py
> **********************************************************************
> File "src/sage/doctest/external.py", line 518, in sage.doctest.external.AvailableSoftware.hidden
> Failed example:
>     available_software.hidden()
> Expected:
>     [...'conway_polynomials',
>      'database_cremona_mini_ellcurve',
>      'database_ellcurves',
>      'database_graphs'...]
> Got:
>     ['database_cremona_mini_ellcurve', 'database_ellcurves', 'database_graphs']
> **********************************************************************
> ```

I can't reproduce this. Could you give us more details and/or share the head of the tree where you see this?

**Comment by vbraun** [comment]
I got this twice in a row when running make distclean && make && make ptestlong

But in general can we make the test maybe less fragile, is it really important that those 4 items have to be consecutively in there? Maybe you just want to check that one is in there so hidden works at all?

**Comment by soehms** [comment]
> is it really important that those 4 items have to be consecutively in there? Maybe you just want to check that one is in there so hidden works at all?

The fact that `conway_polynomials` is missing from the answer is indeed strange. If this would happen when a developer runs `sage -tp --hide=conway_polynomials` we would give him a wrong feedback.

**Comment by soehms** [comment]
The current commit addresses #37905 in the context of this PR. As mentioned in https://github.com/sagemath/sage/pull/37857#issuecomment-2095267529 the connection betwee the content of the variable `_seen` and the result of the method `__contains__` turns out to be problematic when using the hide option since this option may cause an available feature to be temporarily unavailable. Therefore, I suggest decoupling the result from the content of `_seen`.

Note that there is no real need to cache the result of `self._features[idx].is_present()` as it is already cached in the `_is_present` method of the Feature class. From this point of view it seems that there is no real need to set `_seen` to -1. @kwankyu, since you introduced this variable in March 2016 (in [this commit](https://github.com/sagemath/sage/commit/60bc96d74c0a29c6a4fb93dbb58c5df744c8f30d)), can you please confirm that I'm right?

Likewise, there is no need to set `_hidden` to -1. I couldn't reproduce the failure that @vbraun got (see https://github.com/sagemath/sage/pull/37737#issuecomment-2048535021), but I suspect this happened because `_hidden` for the feature `conway_polynomials` was set to -1 (by a thread calling `__contains__` as the feature was not hidden). Therefore, there is a good chance that this failure will no longer occur.

**Comment by mkoeppe** [comment]
test-mod shows this failure. This is a separate test of the modularized distribution sagemath-categories.
```
[sagemath_categories-10.4.beta5] [spkg-check] sage -t --random-seed=207122778768807178970207259884772121442 --environment=sage.all__sagemath_categories sage/doctest/external.py
[sagemath_categories-10.4.beta5] [spkg-check] **********************************************************************
[sagemath_categories-10.4.beta5] [spkg-check] File "sage/doctest/external.py", line 519, in sage.doctest.external.AvailableSoftware.hidden
[sagemath_categories-10.4.beta5] [spkg-check] Failed example:
[sagemath_categories-10.4.beta5] [spkg-check]     sorted(available_software.hidden())
[sagemath_categories-10.4.beta5] [spkg-check] Expected:
[sagemath_categories-10.4.beta5] [spkg-check]     [...'conway_polynomials',...
[sagemath_categories-10.4.beta5] [spkg-check]      'database_cremona_mini_ellcurve',...
[sagemath_categories-10.4.beta5] [spkg-check]      'database_ellcurves',...
[sagemath_categories-10.4.beta5] [spkg-check]      'database_graphs'...]
[sagemath_categories-10.4.beta5] [spkg-check] Got:
[sagemath_categories-10.4.beta5] [spkg-check]     []
[sagemath_categories-10.4.beta5] [spkg-check] **********************************************************************
```

File: src/sage/doctest/external.py [hunk 1]
```diff
@@ -495,5 +503,28 @@ def seen(self):
                 for feature, seen in zip(self._features, self._seen)
                 if seen > 0]
 
+    def hidden(self):
+        """
+        Return the list of detected hidden external software.
+
+        EXAMPLES::
+
+            sage: from sage.doctest.external import available_software
```
Comment (mkoeppe) [inline src/sage/doctest/external.py:513]:
```suggestion
            sage: # needs conway_polynomials database_cremona_mini_ellcurve database_ellcurves database_graphs
            sage: from sage.doctest.external import available_software
```

Comment (soehms) [reply src/sage/doctest/external.py:513]: Sorry, I did it by myself, because I missed that I just need to commit your suggestion! Thanks!

**Comment by soehms** [comment]
Thanks!

**Pull request merged**